src/core/qgsexpression.cpp

```diff
@@ -4225,6 +4230,16 @@ QString QgsExpression::quotedValue( const QVariant &value, QVariant::Type type )
     case QVariant::Bool:
       return value.toBool() ? "TRUE" : "FALSE";
 
+    case QVariant::List:
+    {
+      QStringList quotedValues;
+      Q_FOREACH ( const QVariant &v, value.toList() )
+      {
+        quotedValues += quotedValue( v );
+      }
+      return QStringLiteral( "array( %1 )" ).arg( quotedValues.join( QStringLiteral( ", " ) ) );
```

Comment:
If you change this to use the char based join it'll be MUCH faster:

    return QStringLiteral( "array( %1 )" ).arg( quotedValues.join( ',' ) ) );